### ISSUE TITLE
BYOL loss function

### ISSUE THREAD

**Issue by Jackie-LJQ** [open-mmlab/mmselfsup#118]
Hello, 
I'm trying to reproduce results of byol on Imagenet dataset. I checked your work and I am confused about the loss function.

I think the loss function should be summation on each pair of image sample vector. Instead of 
https://github.com/open-mmlab/OpenSelfSup/blob/1db69ecebbc129e8fa90cdcea6f2082f0a4e3d17/openselfsup/models/heads/latent_pred_head.py#L34
I think it should be `loss = 2-2 * (pred_norm * target_norm).sum(dim=-1)`

In byol paper the loss function was primary described as squared error between two vectors from online model and target model, and the squared error could be simplified as something related to summation of element-wise multiplication.
![123](https://user-images.githubusercontent.com/57075746/123394250-815a6c00-d5d1-11eb-91a3-62cae9392ec9.png)
Also in the appendix J.3 in the origin paper, the code they provided uses sum(-1)
![loss](https://user-images.githubusercontent.com/57075746/123394773-15c4ce80-d5d2-11eb-896e-15a9178d06a0.png)

Why you use summation over all elements (sum())?

Though I didn't actually run your code due to the environment issue, I tried translate the core part of your work and merged to my code. The loss log shows like below. Is this similar to your log?
![log](https://user-images.githubusercontent.com/57075746/123395405-bf0bc480-d5d2-11eb-97cd-cd4a6d0b59a2.png)

**Comment by Jiahao000** [comment]
The two implementations are equivalent. Please check carefully. The summation is then averaged in [latent_pred_head.py L36](https://github.com/open-mmlab/OpenSelfSup/blob/master/openselfsup/models/heads/latent_pred_head.py#L36). This is equivalent to `.sum(dim=-1).mean()`. The absolute loss value in your log is too large, you should average the loss within a batch.

**Comment by Jackie-LJQ** [comment]
I understand .sum() equals to .sum(dim=-1).mean(), but I thought they are different from   (2-(p*t).sum(dim=-1)).mean().
Though I just found the code provided in paper use -2*(p*t) instead of 2-(p*t)...

I used lossfunc = (2-(p*t).sum(dim=-1)).mean() in my own implementation. In my opinion this function matched the math function in paper. During training though it seems that the loss in log decreased, from 1.7 to 0.3. The linear classification results is a mess, like the backbone is random. 

I will tried -2*(p*t).sum(). later

BTW Do you think using (2-(p*t).sum(dim=-1)).mean() as loss function is wrong. I got top1 acc equal to 0.1...

Thx

**Comment by Jiahao000** [comment]
It seems that you are missing a coefficient before (p*t). It should be `2 - 2 * (p * t).sum(dim=-1).mean()`.

**Comment by Jackie-LJQ** [comment]
I'll tried them. Thx

**Comment by Jackie-LJQ** [comment]
I tried use `loss_func = 2 - 2 * (p * t).sum(dim=-1).mean()` and `loss_func = -2 * (p * t).sum()`. Both gave good results comparing to my old implementation using `loss_func=2 -  (p * t).sum(dim=-1).mean()`. It seems like the coefficient make great difference. After 200 epochs pretrain on Imagenet1k dataset, the first loss function gave top1 Acc up to 60 and the second gave the best top1 Acc equals to 64. Hopefully with more careful design I could reach the performance proposed by in paper. Thanks for your suggestion.

Also I found the learning rate in the downstream classification task takes great effect on the result which makes it hard to do a fair comparison. It's hard to tell whether the good results comes from the downstream learning rate or the method used in pretrain.

Do you have plan to implement mocov3 based on Resnet50 backbone? I have tried it but the result didn't reach the paper's.

**Comment by Jiahao000** [comment]
We currently do not support this variant. We will appreciate it if you would like to contribute.
